go-controller/pkg/node/controllers/egressip/pod.go

```diff
@@ -0,0 +1,277 @@
+)
+
+// podIPConfig holds pod specific info to implement egress IP for non-OVN managed networks for a single pod IP. A pod may
+// contain multiple IPs (one for single stack, 2 for dual stack).
+type podIPConfig struct {
+	failed      bool // used for retry
+	v6          bool
+	ipTableRule iptables.RuleArg
+	ipRule      netlink.Rule
+}
+
+func newPodIPConfig() *podIPConfig {
+	return &podIPConfig{}
+}
+
+func (pIC *podIPConfig) equal(pIC2 *podIPConfig) bool {
+	if pIC.v6 != pIC2.v6 {
+		return false
+	}
+	if !equal(pIC.ipTableRule.Args, pIC2.ipTableRule.Args) {
+		return false
+	}
+	if pIC.ipRule.String() != pIC2.ipRule.String() {
+		return false
+	}
+	return true
+}
+
+// podIPConfigList holds a list of podIPConfig to configure EIP for a single pod and its IPs.
+// Each item in the list represents one pod IP.
+type podIPConfigList struct {
+	elems []*podIPConfig
+}
+
+func newPodIPConfigList() *podIPConfigList {
+	return &podIPConfigList{elems: []*podIPConfig{}}
+}
+
+func (pICL *podIPConfigList) Len() int {
+	return len(pICL.elems)
+}
+
+func (pICL *podIPConfigList) hasWithoutError(pIC2 *podIPConfig) bool {
+	for _, pIC := range pICL.elems {
+		if pIC.equal(pIC2) && !pIC.failed {
+			return true
+		}
+	}
+	return false
+}
+
+func (pICL *podIPConfigList) has(pIC2 *podIPConfig) bool {
+	for _, pIC := range pICL.elems {
+		if pIC.equal(pIC2) {
+			return true
+		}
+	}
+	return false
+}
+
+func (pICL *podIPConfigList) remove(idxs ...int) {
+	if len(idxs) == 0 {
+		return
+	}
+	newElems := make([]*podIPConfig, 0, len(pICL.elems))
+	idxToDelete := sets.New[int](idxs...)
+
+	for existingIdx, pIC := range pICL.elems {
+		if !idxToDelete.Has(existingIdx) {
+			newElems = append(newElems, pIC)
+		}
+	}
+	pICL.elems = newElems
+}
+
+// InsertOverwriteFailed should be used to add elements to the podIPConfigList that have failed to be applied.
+func (pICL *podIPConfigList) InsertOverwriteFailed(pICs ...podIPConfig) {
+	for _, pIC := range pICs {
+		pIC.failed = true
+		pICL.insertOverwrite(pIC)
+	}
+}
+
+// Insert should be used to add elements to the podIPConfigList
+func (pICL *podIPConfigList) Insert(pICs ...podIPConfig) {
+	for _, pc := range pICs {
+		pICL.insertOverwrite(pc)
+	}
+}
+
+func (pICL *podIPConfigList) insertOverwrite(pIC podIPConfig) {
+	var dupeIdxs []int
+	for idx, existingPIC := range pICL.elems {
+		if existingPIC.equal(&pIC) {
+			dupeIdxs = append(dupeIdxs, idx)
+		}
+	}
+	pICL.remove(dupeIdxs...)
+	pICL.elems = append(pICL.elems, &pIC)
+}
+
+func (c *Controller) onPodAdd(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expecting %T but received %T", &corev1.Pod{}, obj))
+		return
+	}
+	if pod == nil {
+		utilruntime.HandleError(errors.New("invalid Pod provided to onPodAdd()"))
+		return
+	}
+	// if the pod does not have IPs AND there are no multus network status annotations found, skip it
+	if len(pod.Status.PodIPs) == 0 && len(pod.Annotations[nettypes.NetworkStatusAnnot]) == 0 {
```

Comment:
I guess we shouldn't bother with the network status annotation here either?

Reply:
No removed.